rust/automerge/tests/text_encoding.rs

```diff
@@ -0,0 +1,625 @@
+            Self::ByEncoding {
+                code_point,
+                utf8,
+                utf16,
+                grapheme,
+            } => match encoding {
+                Encoding::UnicodeCodePoint => {
+                    assert_eq!(actual, code_point, "failed for {}", encoding)
+                }
+                Encoding::Utf8CodeUnit => assert_eq!(actual, utf8, "failed for {}", encoding),
+                Encoding::Utf16CodeUnit => assert_eq!(actual, utf16, "failed for {}", encoding),
+                Encoding::GraphemeCluster => {
+                    assert_eq!(actual, grapheme, "failed for {}", encoding)
+                }
+            },
+        }
+    }
+}
+
+struct Scenario<F, T> {
+    text: &'static str,
+    action: F,
+    expected: Expected<T>,
+}
+
+impl<F: Fn(&mut AutoCommit, &automerge::ObjId, Encoding) -> T, T: PartialEq + std::fmt::Debug>
+    Scenario<F, T>
+{
+    fn run(&self) {
+        for encoding in [
+            Encoding::UnicodeCodePoint,
+            Encoding::Utf8CodeUnit,
+            Encoding::Utf16CodeUnit,
+            Encoding::GraphemeCluster,
+        ] {
+            self.run_with_encoding(encoding);
+        }
+    }
+
+    fn run_with_encoding(&self, encoding: Encoding) {
+        let mut doc = AutoCommit::new_with_encoding(encoding.into());
+        let text = doc.put_object(ROOT, "text", ObjType::Text).unwrap();
+        doc.splice_text(&text, 0, 0, self.text).unwrap();
+        let result = (self.action)(&mut doc, &text, encoding);
+        self.expected.assert(&result, encoding);
+    }
+}
+
+impl<
+        F: Fn(&mut AutoCommit, &automerge::ObjId, Encoding) -> Result<T, String>,
+        T: PartialEq + std::fmt::Debug,
+    > Scenario<F, T>
+{
+    fn run_fallible(&self) {
+        for encoding in [
+            Encoding::UnicodeCodePoint,
+            Encoding::Utf8CodeUnit,
+            Encoding::Utf16CodeUnit,
+            Encoding::GraphemeCluster,
+        ] {
+            self.run_fallible_with_encoding(encoding);
+        }
+    }
+
+    fn run_fallible_with_encoding(&self, encoding: Encoding) {
+        let mut doc = AutoCommit::new_with_encoding(encoding.into());
+        let text = doc.put_object(ROOT, "text", ObjType::Text).unwrap();
+        doc.splice_text(&text, 0, 0, self.text).unwrap();
+        let result = (self.action)(&mut doc, &text, encoding);
+        match result {
+            Ok(result) => self.expected.assert(&result, encoding),
+            Err(e) => panic!("failed for {}: {}", encoding, e),
+        }
+    }
+}
+
+#[test]
+fn length() {
+    Scenario {
+        text: "hello🫠",
+        action: |doc: &mut AutoCommit, text: &ObjId, _encoding: Encoding| doc.length(text),
+        expected: Expected::ByEncoding {
+            code_point: 6,
+            utf8: 9,
+            utf16: 7,
+            grapheme: 6,
+        },
+    }
+    .run();
+}
+
+#[test]
+fn splice_text() {
+    Scenario {
+        text: "hello 🫠 world",
```

Comment:
mmm... could you repeat the test using `🧑‍🧑‍🧒‍🧒` instead of that emoji?

**Insertion** is always based on `rust::chars` [**[ref](https://github.com/automerge/automerge/blob/a3185700b3400d9f00060ee1dcaeb8675cbab1db/rust/automerge/src/transaction/inner.rs#L601)**], which splice text based on Unicode scalars. Probably it is not expected, and it seems tests are passing because the picked emoji is 1:1 between Unicode-scalars and grapheme clusters.